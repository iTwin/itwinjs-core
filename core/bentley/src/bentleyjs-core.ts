/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
export * from "./Assert";
export * from "./AsyncMutex";
export * from "./BeEvent";
export * from "./BeSQLite";
export * from "./BentleyError";
export * from "./BentleyLoggerCategory";
export * from "./ByteStream";
export * from "./ClientRequestContext";
export * from "./Compare";
export * from "./CompressedId64Set";
export * from "./Config";
export * from "./Dictionary";
export * from "./Disposable";
export * from "./ElectronUtils";
export * from "./Id";
export * from "./IndexMap";
export * from "./InstanceOf";
export * from "./JsonUtils";
export * from "./LRUMap";
export * from "./Logger";
export * from "./ObservableSet";
export * from "./OneAtATimeAction";
export * from "./OrderedId64Iterable";
export * from "./partitionArray";
export * from "./PriorityQueue";
export * from "./SortedArray";
export * from "./StringUtils";
export * from "./Time";

/** @packageDocumentation
 * @module Utils
 */

/** @docs-package-description
 * The bentleyjs-core package contains classes to solve problems that are common for both client and server use cases.
 */
/**
 * @docs-group-description BeSQLite
 * Classes for working with SQLite databases. SQLite underlies IModelDb and ECDb - see [Executing ECSQL]($docs/learning/ECSQL.md)
 */
/**
 * @docs-group-description Configuration
 * Class for easily managing configuration variables for an iModel.js application.
 */
/**
 * @docs-group-description Errors
 * Classes for working with errors.
 */
/**
 * @docs-group-description Events
 * Classes for raising and handling events.
 */
/**
 * @docs-group-description Ids
 * Classes for working with unique identifiers.
 */
/**
 * @docs-group-description Logging
 * Classes for configuring and logging diagnostic messages - see [Learning about Logging]($docs/learning/common/Logging.md)
 */
/**
 * @docs-group-description Collections
 * Specialized, customizable collection classes like priority queues.
 */
/**
 * @docs-group-description Utils
 * Miscellaneous utility classes.
 */
