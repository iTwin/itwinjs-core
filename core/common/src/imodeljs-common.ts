/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
export * from "./AmbientOcclusion";
export * from "./AnalysisStyle";
export * from "./BackgroundMapSettings";
export * from "./BriefcaseTypes";
export * from "./Camera";
export * from "./ChangedElements";
export * from "./ClipStyle";
export * from "./CloudStorage";
export * from "./CloudStorageTileCache";
export * from "./Code";
export * from "./ColorByName";
export * from "./ColorDef";
export * from "./CommonLoggerCategory";
export * from "./DisplayStyleSettings";
export * from "./domains/FunctionalElementProps";
export * from "./domains/GenericElementProps";
export * from "./ECSqlTypes";
export * from "./ElementProps";
export * from "./EntityProps";
export * from "./FeatureGates";
export * from "./FeatureIndex";
export * from "./FeatureSymbology";
export * from "./FeatureTable";
export * from "./Fonts";
export * from "./Frustum";
export * from "./GeoCoordinateServices";
export * from "./geometry/AreaPattern";
export * from "./geometry/BoundingSphere";
export * from "./geometry/Cartographic";
export * from "./geometry/ElementGeometry";
export * from "./geometry/FrustumPlanes";
export * from "./geometry/GeometryStream";
export * from "./geometry/ImageGraphic";
export * from "./geometry/LineStyle";
export * from "./geometry/Placement";
export * from "./geometry/TextString";
export * from "./GeometryContainment";
export * from "./GeometryParams";
export * from "./GeometrySummary";
export * from "./Gradient";
export * from "./GraphicParams";
export * from "./GroundPlane";
export * from "./HiddenLine";
export * from "./Hilite";
export * from "./HSLColor";
export * from "./HSVColor";
export * from "./Image";
export * from "./IModel";
export * from "./IModelError";
export * from "./IModelVersion";
export * from "./ipc/BackendIpc";
export * from "./ipc/FrontendIpc";
export * from "./ipc/IpcSocket";
export * from "./ipc/NativeAppIpc";
export * from "./ipc/IpcWebSocket";
export * from "./LightSettings";
export * from "./LinePixels";
export * from "./MapImagerySettings";
export * from "./MapLayerSettings";
export * from "./MassProperties";
export * from "./MaterialProps";
export * from "./ModelClipGroup";
export * from "./ModelProps";
export * from "./OctEncodedNormal";
export * from "./Paging";
export * from "./PlanProjectionSettings";
export * from "./QPoint";
export * from "./Render";
export * from "./RenderMaterial";
export * from "./RenderSchedule";
export * from "./RenderTexture";
export * from "./RgbColor";
export * from "./RpcInterface";
export * from "./RpcManager";
export * from "./SkyBox";
export * from "./Snapping";
export * from "./SolarCalculate";
export * from "./SolarShadows";
export * from "./SpatialClassificationProps";
export * from "./SubCategoryAppearance";
export * from "./SubCategoryOverride";
export * from "./TerrainSettings";
export * from "./TextureMapping";
export * from "./TextureProps";
export * from "./ThematicDisplay";
export * from "./Thumbnail";
export * from "./TileProps";
export * from "./Tween";
export * from "./ViewDetails";
export * from "./ViewFlags";
export * from "./ViewProps";
export * from "./oidc/DesktopAuthorizationClientConfiguration";
export * from "./oidc/MobileAuthorizationClientConfiguration";
export * from "./rpc/DevToolsRpcInterface";
export * from "./rpc/EditorRpcInterface";
export * from "./rpc/IModelReadRpcInterface";
export * from "./rpc/IModelTileRpcInterface";
export * from "./rpc/IModelWriteRpcInterface";
export * from "./rpc/ModelGeometryChanges";
export * from "./rpc/SnapshotIModelRpcInterface";
export * from "./rpc/StandaloneIModelRpcInterface";
export * from "./rpc/TestRpcManager";
export * from "./rpc/WipRpcInterface";
export * from "./rpc/core/RpcConfiguration";
export * from "./rpc/core/RpcConstants";
export * from "./rpc/core/RpcControl";
export * from "./rpc/core/RpcInvocation";
export * from "./rpc/core/RpcMarshaling";
export * from "./rpc/core/RpcOperation";
export * from "./rpc/core/RpcPendingQueue";
export * from "./rpc/core/RpcProtocol";
export * from "./rpc/core/RpcRegistry";
export * from "./rpc/core/RpcRequest";
export * from "./rpc/core/RpcRequestContext";
export * from "./rpc/core/RpcRoutingToken";
export * from "./rpc/core/RpcPush";
export * from "./rpc/mobile/MobileRpcManager";
export * from "./rpc/mobile/MobileRpcManager";
export * from "./rpc/mobile/MobileRpcProtocol";
export * from "./rpc/mobile/MobileRpcRequest";
export * from "./rpc/web/BentleyCloudRpcManager";
export * from "./rpc/web/BentleyCloudRpcProtocol";
export * from "./rpc/web/OpenAPI";
export * from "./rpc/web/RpcMultipart";
export * from "./rpc/web/WebAppRpcProtocol";
export * from "./rpc/web/WebAppRpcRequest";
export * from "./tile/B3dmTileIO";
export * from "./tile/CompositeTileIO";
export * from "./tile/GltfTileIO";
export * from "./tile/I3dmTileIO";
export * from "./tile/IModelTileIO";
export * from "./tile/PntsTileIO";
export * from "./tile/TileIO";
export * from "./tile/TileMetadata";

/** @docs-package-description
 * The imodeljs-common package contains classes for working with iModels that can be used in both [frontend]($docs/learning/frontend/index.md) and [backend]($docs/learning/backend/index.md).
 */
/**
 * @docs-group-description Entities
 * Definitions of the "props" interfaces and types that define the [wire format]($docs/learning/wireformat.md) for communication between the frontend and backend about entities (models, elements, etc) contained in an iModel.
 */
/**
 * @docs-group-description Codes
 * Types for working with [Codes]($docs/bis/intro/codes.md).
 */
/**
 * @docs-group-description Geometry
 * Types for working with geometry.
 */
/**
 * @docs-group-description Serialization
 * Types for serializing geometry
 */
/**
 * @docs-group-description Views
 * Types for defining graphical views of the contents of an iModel.
 */
/**
 * @docs-group-description DisplayStyles
 * Types for describing how the contents of Views should be rendered.
 */
/**
 * @docs-group-description Rendering
 * Types describing geometry, views, and symbology for consumption by a display system.
 */
/**
 * @docs-group-description Symbology
 * Types that define the appearance of geometry.
 */
/**
 * @docs-group-description iModels
 * Types for working with [iModels]($docs/learning/IModels.md) in both the frontend and backend.
 */
/**
 * @docs-group-description RpcInterface
 * Types for working with [RpcInterfaces]($docs/learning/RpcInterface.md).
 */
/**
 * @docs-group-description IpcSocket
 * Types for working with [IpcInterfaces]($docs/learning/IpcInterface.md).
 */
/**
 * @docs-group-description ECSQL
 * Types for working with [ECSQL]($docs/learning/ECSQL.md), [Spatial Queries]($docs/learning/SpatialQueries.md), and [ECSQL Geometry Functions]($docs/learning/GeometrySqlFuncs.md).
 */
/**
 * @docs-group-description Logging
 * Logger categories used by this package.
 */
/**
 * @docs-group-description CloudStorage
 * Types for working with Cloud Storage.
 */
/**
 * @docs-group-description Tween
 * Tweening library adapted from tween.js.
 */
/**
 * @docs-group-description Tile
 * Types for working with 3d tile formats.
 */
/**
 * @docs-group-description Utils
 * Miscellaneous utility classes.
 */
/**
 * @docs-group-description NativeApp
 * [Native applications]($docs/learning/NativeApps.md)
 */
