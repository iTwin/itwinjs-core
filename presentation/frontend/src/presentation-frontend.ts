/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/**
 * @module Core
 *
 * @docs-group-description Core
 * Common types used for retrieving presentation data from iModels.
 */
export { Presentation } from "./presentation-frontend/Presentation";
export {
  IModelContentChangeEventArgs, IModelHierarchyChangeEventArgs, PresentationManager, PresentationManagerProps,
} from "./presentation-frontend/PresentationManager";
export { RulesetManager } from "./presentation-frontend/RulesetManager";
export { RulesetVariablesManager } from "./presentation-frontend/RulesetVariablesManager";
export {
  FavoritePropertiesManager, PropertyFullName, FavoritePropertiesOrderInfo,
  FavoritePropertiesScope, getFieldInfos, createFieldOrderInfos,
} from "./presentation-frontend/favorite-properties/FavoritePropertiesManager";
export { IFavoritePropertiesStorage } from "./presentation-frontend/favorite-properties/FavoritePropertiesStorage";
export { NodeIdentifier, StateTracker } from "./presentation-frontend/StateTracker";

/**
 * @module Logging
 *
 * @docs-group-description Logging
 * Types related to logging in this package.
 */
export * from "./presentation-frontend/FrontendLoggerCategory";

/**
 * @module UnifiedSelection
 *
 * @docs-group-description UnifiedSelection
 * Types related to [unified selection]($docs/learning/presentation/Unified-Selection/index.md).
 */
export {
  SelectionChangeEvent, SelectionChangeEventArgs, SelectionChangeType, SelectionChangesListener,
} from "./presentation-frontend/selection/SelectionChangeEvent";
export { ISelectionProvider } from "./presentation-frontend/selection/ISelectionProvider";
export { SelectionManager, SelectionManagerProps } from "./presentation-frontend/selection/SelectionManager";
export { SelectionScopesManager, SelectionScopesManagerProps, getScopeId } from "./presentation-frontend/selection/SelectionScopesManager";
export { SelectionHandler, SelectionHandlerProps } from "./presentation-frontend/selection/SelectionHandler";
export { HiliteSet, HiliteSetProvider, HiliteSetProviderProps } from "./presentation-frontend/selection/HiliteSetProvider";
export { SelectionHelper } from "./presentation-frontend/selection/SelectionHelper";
