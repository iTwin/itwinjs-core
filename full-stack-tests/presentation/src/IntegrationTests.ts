/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import * as chai from "chai";
import chaiSubset from "chai-subset";
import * as cpx from "cpx";
import * as fs from "fs";
import * as path from "path";
import sinonChai from "sinon-chai";
import { ClientRequestContext, Logger, LogLevel } from "@bentley/bentleyjs-core";
import { loadEnv } from "@bentley/config-loader";
import { IModelAppOptions, NoRenderApp } from "@bentley/imodeljs-frontend";
import { I18NOptions } from "@bentley/imodeljs-i18n";
import { TestUsers } from "@bentley/oidc-signin-tool/lib/TestUsers";
import { TestUtility } from "@bentley/oidc-signin-tool/lib/TestUtility";
import {
  HierarchyCacheMode, Presentation as PresentationBackend, PresentationBackendLoggerCategory, PresentationBackendNativeLoggerCategory,
  PresentationProps as PresentationBackendProps,
} from "@bentley/presentation-backend";
import { RequestPriority } from "@bentley/presentation-common";
import { PresentationComponentsLoggerCategory } from "@bentley/presentation-components";
import { PresentationFrontendLoggerCategory, PresentationManagerProps as PresentationFrontendProps } from "@bentley/presentation-frontend";
import { initialize as initializeTesting, PresentationTestingInitProps, terminate as terminateTesting } from "@bentley/presentation-testing";

chai.use(sinonChai);
chai.use(chaiSubset);

loadEnv(path.join(__dirname, "..", ".env"));

const copyBentleyBackendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "assets");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

const copyBentleyFrontendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "public");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

class IntegrationTestsApp extends NoRenderApp {
  protected static supplyI18NOptions(): I18NOptions {
    const urlTemplate = `file://${path.join(path.resolve("lib/public/locales"), "{{lng}}/{{ns}}.json").replace(/\\/g, "/")}`;
    return { urlTemplate };
  }

  public static async startup(opts?: IModelAppOptions): Promise<void> {
    await NoRenderApp.startup({ ...opts, i18n: this.supplyI18NOptions() });
    cpx.copySync(`assets/**/*`, "lib/assets");
    copyBentleyBackendAssets("lib/assets");
    copyBentleyFrontendAssets("lib/public");
  }
}

const initializeCommon = async (props: { backendTimeout?: number, useClientServices?: boolean }) => {
  // init logging
  Logger.initializeToConsole();
  Logger.setLevelDefault(LogLevel.Warning);
  Logger.setLevel(PresentationBackendNativeLoggerCategory.ECObjects, LogLevel.Warning);
  Logger.setLevel(PresentationBackendNativeLoggerCategory.ECPresentation, LogLevel.Info);
  Logger.setLevel(PresentationBackendNativeLoggerCategory.ECPresentation_Localization, LogLevel.Info);
  Logger.setLevel(PresentationBackendLoggerCategory.Package, LogLevel.Info);
  Logger.setLevel(PresentationFrontendLoggerCategory.Package, LogLevel.Info);
  Logger.setLevel(PresentationComponentsLoggerCategory.Package, LogLevel.Info);

  const libDir = path.resolve("lib");
  console.log(`Backend's lib directory path: ${libDir}`); // eslint-disable-line no-console

  const backendInitProps: PresentationBackendProps = {
    requestTimeout: props.backendTimeout ?? 0,
    rulesetDirectories: [path.join(libDir, "assets", "rulesets")],
    localeDirectories: [path.join(libDir, "assets", "locales")],
    activeLocale: "en-PSEUDO",
    taskAllocationsMap: {
      [RequestPriority.Max]: 1,
    },
    cacheConfig: { mode: HierarchyCacheMode.Disk, directory: path.join(libDir, "cache") },
  };
  const frontendInitProps: PresentationFrontendProps = {
    activeLocale: "en-PSEUDO",
  };

  const frontendAppOptions: IModelAppOptions = {
    authorizationClient: props.useClientServices
      ? TestUtility.getAuthorizationClient(TestUsers.regular)
      : undefined,
  };

  if (props.useClientServices)
    await frontendAppOptions.authorizationClient!.signIn(new ClientRequestContext());

  const presentationTestingInitProps: PresentationTestingInitProps = {
    backendProps: backendInitProps,
    frontendProps: frontendInitProps,
    frontendApp: IntegrationTestsApp,
    frontendAppOptions,
  };

  await initializeTesting(presentationTestingInitProps);
  console.log(`Backend PID: ${process.pid}`); // eslint-disable-line no-console
};

export const initialize = async (backendTimeout: number = 0) => {
  await initializeCommon({ backendTimeout });
};

export const initializeWithClientServices = async () => {
  await initializeCommon({ useClientServices: true });
};

export const terminate = async () => {
  await terminateTesting();
};

export const resetBackend = () => {
  const props = PresentationBackend.initProps;
  PresentationBackend.terminate();
  PresentationBackend.initialize(props);
};
