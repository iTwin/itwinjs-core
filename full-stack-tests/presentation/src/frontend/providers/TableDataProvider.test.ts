/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as sinon from "sinon";
import { Id64 } from "@bentley/bentleyjs-core";
import { ModelProps } from "@bentley/imodeljs-common";
import { IModelConnection, SnapshotConnection } from "@bentley/imodeljs-frontend";
import {
  ContentSpecificationTypes, InstanceKey, KeySet, RelationshipDirection, RelationshipMeaning, Ruleset, RuleTypes,
} from "@bentley/presentation-common";
import { PresentationTableDataProvider } from "@bentley/presentation-components";
import { Presentation } from "@bentley/presentation-frontend";
import { PropertyRecord, PropertyValueFormat } from "@bentley/ui-abstract";
import { SortDirection } from "@bentley/ui-core";
import { initialize, terminate } from "../../IntegrationTests";

const RULESET: Ruleset = {
  id: "localization test",
  rules: [{
    ruleType: RuleTypes.Content,
    specifications: [{
      specType: ContentSpecificationTypes.SelectedNodeInstances,
    }],
  }],
};

interface MeaningfulInstances {
  repositoryModel: ModelProps;
  dictionaryModel: ModelProps;
  physicalModel: ModelProps;
}
const createMeaningfulInstances = async (imodel: IModelConnection): Promise<MeaningfulInstances> => {
  return {
    repositoryModel: (await imodel.models.queryProps({ from: "bis.RepositoryModel" }))[0],
    dictionaryModel: (await imodel.models.queryProps({ from: "bis.DictionaryModel", wantPrivate: true }))[0],
    physicalModel: (await imodel.models.queryProps({ from: "bis.PhysicalModel" }))[0],
  };
};

describe("TableDataProvider", async () => {

  let imodel: IModelConnection;
  let instances: MeaningfulInstances;
  let provider: PresentationTableDataProvider;

  before(async () => {
    await initialize();
    const testIModelName: string = "assets/datasets/Properties_60InstancesWithUrl2.ibim";
    imodel = await SnapshotConnection.openFile(testIModelName);
    instances = await createMeaningfulInstances(imodel);
  });

  beforeEach(async () => {
    provider = new PresentationTableDataProvider({ imodel, ruleset: RULESET, pageSize: 10 });
  });

  after(async () => {
    await imodel.close();
    await terminate();
  });

  describe("getColumns", () => {

    it("returns columns for a single instance", async () => {
      provider.keys = new KeySet([instances.physicalModel]);
      const columns = await provider.getColumns();
      expect(columns).to.matchSnapshot();
    });

    it("returns columns for multiple instances", async () => {
      provider.keys = new KeySet([instances.repositoryModel, instances.physicalModel]);
      const columns = await provider.getColumns();
      expect(columns).to.matchSnapshot();
    });

    it("returns extracted columns from instances which have SameInstance relationshipMeaning", async () => {
      provider.keys = new KeySet([{ className: "Generic:PhysicalObject", id: "0x74" }]);
      const columns = await provider.getColumns();
      expect(columns.length).to.eq(32);
    });

  });

  describe("getRowsCount", () => {

    it("returns total number of instances when less than page size", async () => {
      provider.keys = new KeySet([instances.repositoryModel, instances.physicalModel]);
      const count = await provider.getRowsCount();
      expect(count).to.eq(2);
    });

    it("returns total number of instances when more than page size", async () => {
      const keys = await imodel.elements.queryProps({ from: "bis.PhysicalElement", limit: 20 });
      provider.keys = new KeySet(keys);
      const count = await provider.getRowsCount();
      expect(count).to.eq(20);
    });

  });

  describe("getRow", () => {

    it("returns first row", async () => {
      provider.keys = new KeySet([instances.physicalModel]);
      const row = await provider.getRow(0);
      expect(row).to.matchSnapshot();
    });

    it("returns row with extracted cells from instances which have SameInstance relationshipMeaning", async () => {
      provider.keys = new KeySet([{ className: "Generic:PhysicalObject", id: "0x74" }]);
      const row = await provider.getRow(0);
      expect(row.cells.length).to.eq(32);
    });

    it("returns row with merged cells from instances which have SameInstance relationshipMeaning and more than one value in it", async () => {
      provider = new PresentationTableDataProvider({
        imodel,
        ruleset: {
          id: "test",
          rules: [{
            ruleType: RuleTypes.Content,
            specifications: [{
              specType: ContentSpecificationTypes.SelectedNodeInstances,
            }],
          }, {
            ruleType: RuleTypes.ContentModifier,
            class: {
              schemaName: "BisCore",
              className: "Model",
            },
            relatedProperties: [
              {
                propertiesSource: {
                  relationship: {
                    schemaName: "BisCore",
                    className: "ModelContainsElements",
                  },
                  direction: RelationshipDirection.Forward,
                },
                handleTargetClassPolymorphically: true,
                relationshipMeaning: RelationshipMeaning.SameInstance,
                properties: [
                  "UserLabel",
                  "CodeValue",
                ],
              },
            ],
          }],
        },
        pageSize: 10,
      });
      provider.keys = new KeySet([instances.physicalModel]);

      const row = await provider.getRow(0);
      expect(row.cells.length).to.eq(5);
      expect(row.cells).to.containSubset([{
        key: "pc_bis_Model_ModeledElement",
        mergedCellsCount: undefined,
      }, {
        key: "pc_bis_Element_UserLabel",
        mergedCellsCount: 2,
        alignment: "center",
        record: {
          property: { displayLabel: "Physical Object" },
          value: {
            valueFormat: PropertyValueFormat.Array,
            itemsTypeName: "Physical Object",
            items: (value: PropertyRecord[]) => value.length === 2,
          },
        },
      }, {
        key: "pc_bis_Element_CodeValue",
        mergedCellsCount: undefined,
        record: {
          value: {
            value: undefined,
          },
        },
      }, {
        key: "pc_bis_Element_UserLabel_2",
        mergedCellsCount: 2,
        alignment: "center",
        record: {
          property: { displayLabel: "TestClass" },
          value: {
            valueFormat: PropertyValueFormat.Array,
            itemsTypeName: "TestClass",
            items: (value: PropertyRecord[]) => value.length === 60,
          },
        },
      }, {
        key: "pc_bis_Element_CodeValue_2",
        mergedCellsCount: undefined,
        record: {
          value: {
            value: undefined,
          },
        },
      }]);
    });

    it("returns undefined when requesting row with invalid index", async () => {
      provider.keys = new KeySet([instances.physicalModel]);
      const row = await provider.getRow(1);
      expect(row).to.be.undefined;
    });

  });

  it("requests backend only once to get first page", async () => {
    const getContentDescriptorSpy = sinon.spy(Presentation.presentation.rpcRequestsHandler, "getContentDescriptor");
    const getPagedContentSpy = sinon.spy(Presentation.presentation.rpcRequestsHandler, "getPagedContent");
    const getPagedContentSetSpy = sinon.spy(Presentation.presentation.rpcRequestsHandler, "getPagedContentSet");
    provider.keys = new KeySet([instances.physicalModel]);
    provider.pagingSize = 10;

    // request count and first page
    const count = await provider.getRowsCount();
    const row = await provider.getRow(0);

    expect(count).to.not.eq(0);
    expect(row).to.not.be.undefined;
    expect(getContentDescriptorSpy).to.not.be.called;
    expect(getPagedContentSpy).to.be.calledOnce;
    expect(getPagedContentSetSpy).to.not.be.called;
  });

  describe("sorting", () => {

    it("sorts instances ascending", async () => {
      // provide keys so that instances by default aren't sorted in either way
      provider.keys = new KeySet([instances.physicalModel, instances.dictionaryModel, instances.repositoryModel]);
      await provider.sort(0, SortDirection.Ascending); // sort by display label (column index = 0)
      const rows = await Promise.all([0, 1, 2].map(async (index: number) => provider.getRow(index)));
      // expected order:
      // "" (repository model)
      // "BisCore.DictionaryModel" (dictionary model)
      // "Properties_60InstancesWithUrl2" (physical model)
      expect(rows).to.matchSnapshot();
    });

    it("sorts instances descending", async () => {
      // provide keys so that instances by default aren't sorted in either way
      provider.keys = new KeySet([instances.physicalModel, instances.dictionaryModel, instances.repositoryModel]);
      await provider.sort(0, SortDirection.Descending); // sort by display label (column index = 0)
      const rows = await Promise.all([0, 1, 2].map(async (index: number) => provider.getRow(index)));
      // expected order:
      // "Properties_60InstancesWithUrl2" (physical model)
      // "BisCore.DictionaryModel" (dictionary model)
      // "" (repository model)
      expect(rows).to.matchSnapshot();
    });

  });

  describe("filtering", () => {

    it("filters instances", async () => {
      provider.keys = new KeySet([instances.physicalModel, instances.dictionaryModel, instances.repositoryModel]);
      const columns = await provider.getColumns();
      provider.filterExpression = `${columns[0].key} = "Properties_60InstancesWithUrl2"`;
      expect(await provider.getRowsCount()).to.eq(1);
      const row = await provider.getRow(0);
      const rowKey = InstanceKey.fromJSON(JSON.parse(row.key));
      expect(rowKey.id).to.eq(Id64.fromString(instances.physicalModel.id!));
    });

  });

});
