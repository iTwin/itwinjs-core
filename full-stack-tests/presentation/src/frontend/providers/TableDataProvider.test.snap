// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDataProvider getColumns returns columns for a single instance 1`] = `
Array [
  Object {
    "editable": false,
    "filterable": false,
    "key": "pc_bis_Model_ModeledElement",
    "label": "Modeled Element",
    "sortable": true,
  },
]
`;

exports[`TableDataProvider getColumns returns columns for multiple instances 1`] = `
Array [
  Object {
    "editable": false,
    "filterable": false,
    "key": "pc_bis_Model_ModeledElement",
    "label": "Modeled Element",
    "sortable": true,
  },
]
`;

exports[`TableDataProvider getRow returns first row 1`] = `
Object {
  "cells": Array [
    Object {
      "key": "pc_bis_Model_ModeledElement",
      "record": PropertyRecord {
        "description": "Properties_60InstancesWithUrl2",
        "isReadonly": true,
        "property": Object {
          "displayLabel": "Modeled Element",
          "name": "pc_bis_Model_ModeledElement",
          "typename": "navigation",
        },
        "value": Object {
          "displayValue": "Properties_60InstancesWithUrl2",
          "value": "0x1c",
          "valueFormat": 0,
        },
      },
    },
  ],
  "key": "{\\"className\\":\\"BisCore:PhysicalModel\\",\\"id\\":\\"0x1c\\"}",
}
`;

exports[`TableDataProvider sorting sorts instances ascending 1`] = `
Array [
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "",
            "value": undefined,
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:RepositoryModel\\",\\"id\\":\\"0x1\\"}",
  },
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "description": "BisCore.DictionaryModel",
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "BisCore.DictionaryModel",
            "value": "0x10",
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:DictionaryModel\\",\\"id\\":\\"0x10\\"}",
  },
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "description": "Properties_60InstancesWithUrl2",
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "Properties_60InstancesWithUrl2",
            "value": "0x1c",
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:PhysicalModel\\",\\"id\\":\\"0x1c\\"}",
  },
]
`;

exports[`TableDataProvider sorting sorts instances descending 1`] = `
Array [
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "description": "Properties_60InstancesWithUrl2",
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "Properties_60InstancesWithUrl2",
            "value": "0x1c",
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:PhysicalModel\\",\\"id\\":\\"0x1c\\"}",
  },
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "description": "BisCore.DictionaryModel",
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "BisCore.DictionaryModel",
            "value": "0x10",
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:DictionaryModel\\",\\"id\\":\\"0x10\\"}",
  },
  Object {
    "cells": Array [
      Object {
        "key": "pc_bis_Model_ModeledElement",
        "record": PropertyRecord {
          "isReadonly": true,
          "property": Object {
            "displayLabel": "Modeled Element",
            "name": "pc_bis_Model_ModeledElement",
            "typename": "navigation",
          },
          "value": Object {
            "displayValue": "",
            "value": undefined,
            "valueFormat": 0,
          },
        },
      },
    ],
    "key": "{\\"className\\":\\"BisCore:RepositoryModel\\",\\"id\\":\\"0x1\\"}",
  },
]
`;
